external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/DirectKafkaInputDStream.scala

```diff
@@ -143,9 +147,14 @@ private[spark] class DirectKafkaInputDStream[K, V](
 
         lagPerPartition.map { case (tp, lag) =>
           val maxRateLimitPerPartition = ppc.maxRatePerPartition(tp)
-          val backpressureRate = Math.round(lag / totalLag.toFloat * rate)
-          tp -> (if (maxRateLimitPerPartition > 0) {
-            Math.min(backpressureRate, maxRateLimitPerPartition)} else backpressureRate)
+          val effectiveRate = if (rate >= 0) rate else backpressureInitialRate
+          val estimateRate = Math.round(lag / totalLag.toFloat * effectiveRate)
+          val backpressureRate =
+            if (estimateRate > maxRateLimitPerPartition && maxRateLimitPerPartition > 0) {
+              maxRateLimitPerPartition
+            }
+            else estimateRate
```

Comment:
http://spark.apache.org/contributing.html
use braces around the else clause unless it's all on one line